crates/chia-consensus/src/check_time_locks.rs

```diff
@@ -0,0 +1,348 @@
+    prev_transaction_block_height: u32,
+    timestamp: u64,
+) -> PyResult<Option<u32>> {
+    let mut removals_hashmap: HashMap<chia_protocol::BytesImpl<32>, CoinRecord> = HashMap::new();
+    for (k, v) in removal_coin_records.iter() {
+        let key_bytes: Bytes32 = k.extract()?;
+        let coin_record: CoinRecord = v.extract()?;
+        removals_hashmap.insert(key_bytes, coin_record);
+    }
+
+    let res = check_time_locks(
+        &removals_hashmap,
+        bundle_conds,
+        prev_transaction_block_height,
+        timestamp,
+    );
+
+    Ok(res.map(std::convert::Into::into))
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::owned_conditions::OwnedSpendConditions;
+    use chia_protocol::Coin;
+    use rstest::rstest;
+    use std::collections::HashMap;
+
+    fn dummy_coin_record(confirmed_block_index: u32, timestamp: u64) -> CoinRecord {
+        CoinRecord {
+            coin: Coin {
+                parent_coin_info: Bytes32::default(),
+                puzzle_hash: Bytes32::default(),
+                amount: 1,
+            },
+            confirmed_block_index,
+            spent_block_index: 0,
+            coinbase: false,
+            timestamp,
+        }
+    }
+
+    fn dummy_spend_conditions(
+        coin_id: Bytes32,
+        height_relative: Option<u32>,
+        seconds_relative: Option<u64>,
+        before_height_relative: Option<u32>,
+        before_seconds_relative: Option<u64>,
+        birth_height: Option<u32>,
+        birth_seconds: Option<u64>,
+    ) -> OwnedSpendConditions {
+        OwnedSpendConditions {
+            coin_id,
+            parent_id: Bytes32::default(),
+            puzzle_hash: Bytes32::default(),
+            coin_amount: 1,
+            height_relative,
+            seconds_relative,
+            before_height_relative,
+            before_seconds_relative,
+            birth_height,
+            birth_seconds,
+            create_coin: vec![],
+            agg_sig_me: vec![],
+            agg_sig_parent: vec![],
+            agg_sig_puzzle: vec![],
+            agg_sig_amount: vec![],
+            agg_sig_puzzle_amount: vec![],
+            agg_sig_parent_amount: vec![],
+            agg_sig_parent_puzzle: vec![],
+            flags: 0,
+            execution_cost: 0,
+            condition_cost: 0,
+        }
+    }
+
+    fn dummy_bundle(spends: Vec<OwnedSpendConditions>) -> OwnedSpendBundleConditions {
+        OwnedSpendBundleConditions {
+            spends,
+            reserve_fee: 0,
+            height_absolute: 0,
+            seconds_absolute: 0,
+            before_height_absolute: None,
+            before_seconds_absolute: None,
+            agg_sig_unsafe: vec![],
+            cost: 0,
+            removal_amount: 0,
+            addition_amount: 0,
+            validated_signature: true,
+            execution_cost: 0,
+            condition_cost: 0,
+        }
+    }
+
+    #[rstest]
+    #[case(0, 0, 10, 1000, None, None, None)]
```

Comment:
these test cases aren't so easy to read. and it's also not so easy to see that all cases are covered. For example, it looks like you don't cover any cases where there are more than one spend.

Can you think of ways to either make these test cases more concise, or maybe make the `rstest` generate all combinations, and define the outcome programmatically? Maybe some test cases could be separate tests, if that makes it simpler

Reply:
if you implement `Default` for `OwnedSpendBundleConditions`, you can probably just take that as the `#[case]` parameter and construct them by just initializing the specific field(s) you need for each test. That would make it much clearer which constraint is being tested. I think you can do something like: `OwnedSpendBundleConditions { seconds_absolute: 100, ..Default::default()}`